drivers/flash/nrf_qspi_nor: check SRAM source using nrfx + DT sram nodes visible for both nrf53 cores

nvlsianpu opened the pull request:
The driver was using CONFIG_SRAM_BASE_ADDRESS as the value used to
recognize whether source bufer is in RAM. This label provide the
base address of the image SRAM, and not the base of actual HW SRAM.

This patch uses sram0 DTS node register address instead which is
proper way to obtain the SRAM beginning address.

fixes #29467

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
drivers/flash/nrf_qspi_nor: determine RAM source using nrfx

The driver was using CONFIG_SRAM_BASE_ADDRESS as the value used to
recognize whether source buffer is in RAM. This label provide the
base address of the image SRAM, and not the base of actual HW SRAM.

This patch uses nrfx_is_in_ram() instead.

fixes #29467

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

ioannisg commented:
Won't work if the physical SRAM is not sram0, but sram1. I think this is the case in nrf5340 NET,

Still fixes most cases, though.

nvlsianpu commented:
^^ will check that

pabigot reviewed (commented):
Looks fine but not approving until the issue of `sram0`, `sram1`, or both is resolved.

kamlaz commented:
I agree with preceding speakers.

ioannisg reviewed (commented):
Just thinking loud @nvlsianpu @pabigot 
This is an nRF driver, for a Cortex-M platform.
SRAM always starts after 0x20000000 - would this be enough to distinguish between core and system ram? Do we need to use DTS?

ioannisg commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -45,6 +45,19 @@ BUILD_ASSERT(INST_0_SCK_FREQUENCY >= (NRF_QSPI_BASE_CLOCK_FREQ / 16),
+#if DT_NODE_EXISTS(sram0)
+#define NRF_SOC_SRAM_BEGIN DT_REG_ADDR(DT_NODELABEL(sram0))
+#elif DT_NODE_EXISTS(sram1)
+#if DT_NODE_EXISTS(sram0_shared)
```
sram0_shared is, effectively, an SRAM partition, not a physical RAM node.
I understand why you use this, though: because nrF5340 NETWORK uses the sram0_shared.
But this is ugly, IMO.

In my opinion the DTS node sram0 should be still defined for the NETWORK core (but not used, of course). It represents physical SoC memory and I don't see why the NETWORK core should not know about it.

So at the end of the day, both cores could have the sram0 symbol defined, and then this code here would only need to check for sram0 and sram1.

nvlsianpu replied:
I added the sram0 node to NETWORK core DTS. Is that what you want?

nvlsianpu committed:
dts/arm/nordic: made sram node visible for each nrf5340 cpus

Made sram0 visible for NETWORK core target. The core might
access sram0 as shared region might be defined inside it.
Made sram1 visible for APPLICATION core target for keeping
consistence.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

ioannisg reviewed (commented):
A comment : for "symmetry"-related reasons, I think the CPU APP DTSI should also define sram1 if it does not, already. It is a recommendation, though, not a must-do; that's because the sram1 is not accessible by the APP core. But it completes the physical HW representation.

@mbolivar-nordic ^^

anangl commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -699,7 +707,7 @@ static int qspi_nor_write(const struct device *dev, off_t addr,
 	if (size < 4U) {
 		res = write_sub_word(dev, addr, src, size);
-	} else if (((uintptr_t)src < CONFIG_SRAM_BASE_ADDRESS)) {
+	} else if ((uintptr_t)src < NRF_SOC_SRAM_BEGIN) {
```
Why not just  use `if (!nrfx_is_in_ram(src))`?

nvlsianpu replied:
:( didn't know that it exist. @ioannisg Would it be OK what is suggested here ^^?

ioannisg replied:
Of course, that's what I suggested in the first place:

>This is an nRF driver, for a Cortex-M platform.
SRAM always starts after 0x20000000 - would this be enough to distinguish between core and system ram? Do we need to use DTS?

ioannisg replied:
However, this does not mean that adding the sram0, sram1 in DTS is wrong. Actually, it might be useful in other occasions. We might have to come back to this - you can leave it out for now.

mbolivar-nordic replied:
> We might have to come back to this - you can leave it out for now.

I'm assuming this means your ping from earlier is no longer relevant, thanks

ioannisg replied:
Actually, it is relevant since @nvlsianpu has kept the dts node definition in the PR

mbolivar-nordic replied:
> Actually, it is relevant since @nvlsianpu has kept the dts node definition in the PR

OK, since you asked, creating nodes in the app core DTSI that are not accessible by the core feels wrong to me.

DTSpec says:

>The reg property describes the address of the device’s resources within the address space defined by its parent bus.

The app core address space doesn't contain anything there, right? So there shouldn't be a DT node there. Just my opinion.

ioannisg replied:
Yeah, that's why i put it as a "recommendation"... I don't mind much (and I realized I created some noise to @nvlsianpu already, about it)

mbolivar-nordic replied:
> Yeah, that's why i put it as a "recommendation"... I don't mind much (and I realized I created some noise to @nvlsianpu already, about it)

I think it should be taken out unless there is a use case for it. But I won't block the PR.

ioannisg commented:
@pabigot quick look?

pabigot reviewed (approved):
This is fine technically.  The second commit message has:

> This patch uses nrfx_is_in_ram() instead.fixes #29467

which could be fixed up.

ioannisg reviewed (commented):
@nvlsianpu why the sram1 is not added in the cpuapp_qkaa.dtsi and cpuapp_qkaa_eng_a.dtsi?
